storage/schemadb/src/batch.rs

```diff
@@ -0,0 +1,254 @@
+                    .with_label_values(&[cf_name])
+                    .observe(*put_size as f64);
+            }
+        }
+        for (cf_name, num_deletes) in &self.num_deletes {
+            APTOS_SCHEMADB_DELETES_SAMPLED
+                .with_label_values(&[cf_name])
+                .inc_by(*num_deletes as u64);
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct SampledBatchStats {
+    inner: Option<BatchStats>,
+}
+
+impl SampledBatchStats {
+    pub fn put(&mut self, cf_name: ColumnFamilyName, size: usize) {
+        if let Some(inner) = self.inner.as_mut() {
+            inner.put(cf_name, size)
+        }
+    }
+
+    pub fn delete(&mut self, cf_name: ColumnFamilyName) {
+        if let Some(inner) = self.inner.as_mut() {
+            inner.delete(cf_name)
+        }
+    }
+
+    pub fn commit(&self) {
+        if let Some(inner) = self.inner.as_ref() {
+            inner.commit()
+        }
+    }
+}
+
+impl Default for SampledBatchStats {
+    fn default() -> Self {
+        const SAMPLING_PCT: usize = 1;
+
+        Self {
+            inner: (rand::random::<usize>() % 100 < SAMPLING_PCT).then_some(Default::default()),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct RawBatch {
+    pub inner: rocksdb::WriteBatch,
+    pub stats: SampledBatchStats,
+}
+
+pub trait IntoRawBatch {
+    fn into_raw_batch(self, db: &DB) -> DbResult<RawBatch>;
+}
+
+impl IntoRawBatch for RawBatch {
+    fn into_raw_batch(self, _db: &DB) -> DbResult<RawBatch> {
+        Ok(self)
+    }
+}
+
+pub trait WriteBatch: IntoRawBatch {
+    fn stats(&mut self) -> &mut SampledBatchStats;
+
+    /// Adds an insert/update operation to the batch.
+    fn put<S: Schema>(&mut self, key: &S::Key, value: &S::Value) -> DbResult<()> {
+        let key = <S::Key as KeyCodec<S>>::encode_key(key)?;
+        let value = <S::Value as ValueCodec<S>>::encode_value(value)?;
+
+        self.stats()
+            .put(S::COLUMN_FAMILY_NAME, key.len() + value.len());
+        self.raw_put(S::COLUMN_FAMILY_NAME, key, value)
+    }
+
+    fn raw_put(&mut self, cf_name: ColumnFamilyName, key: Vec<u8>, value: Vec<u8>) -> DbResult<()>;
+
+    /// Adds a delete operation to the batch.
+    fn delete<S: Schema>(&mut self, key: &S::Key) -> DbResult<()> {
+        let key = <S::Key as KeyCodec<S>>::encode_key(key)?;
+
+        self.stats().delete(S::COLUMN_FAMILY_NAME);
+        self.raw_delete(S::COLUMN_FAMILY_NAME, key)
+    }
+
+    fn raw_delete(&mut self, cf_name: ColumnFamilyName, key: Vec<u8>) -> DbResult<()>;
+}
+
+#[derive(Debug)]
+pub enum WriteOp {
+    Value { key: Vec<u8>, value: Vec<u8> },
+    Deletion { key: Vec<u8> },
+}
+
+/// `SchemaBatch` holds a collection of updates that can be applied to a DB atomically. The updates
+/// will be applied in the order in which they are added to the `SchemaBatch`.
+#[derive(Debug, Default)]
+pub struct SchemaBatch {
+    rows: DropHelper<HashMap<ColumnFamilyName, Vec<WriteOp>>>,
```

Comment:
could you help me understand why adding this DropHelper for the hashmap? the HashMap should be a simple ds to drop.